src/utils/async-pool.js

```diff
@@ -0,0 +1,193 @@
+     */
+    enqueue(element, priority) {
+        this.heap.push({ element, priority });
+        let current = this.heap.length - 1;
+
+        while (current > 0 && this.heap[current].priority < this.heap[this._getParentIndex(current)].priority) {
+            this._swap(current, this._getParentIndex(current));
+            current = this._getParentIndex(current);
+        }
+    }
+
+    /**
+     * The function returns the element with the highest priority and removes it from the heap.
+     * @returns {any} The function to be executed that returns a promise
+     */
+    dequeue() {
+        if (this.heap.length === 0) {
+            return null;
+        }
+
+        const element = this.heap[0].element;
+        this.heap[0] = this.heap[this.heap.length - 1];
+        this.heap.pop();
+
+        let current = 0;
+        let leftChild = this._getLeftChildIndex(current);
+        let rightChild = this._getRightChildIndex(current);
+
+        while ((leftChild < this.heap.length && this.heap[current].priority > this.heap[leftChild].priority) ||
+                (rightChild < this.heap.length && this.heap[current].priority > this.heap[rightChild].priority)) {
+
+            let minChild = leftChild;
+            if (rightChild < this.heap.length && this.heap[rightChild].priority < this.heap[leftChild].priority) {
+                minChild = rightChild;
+            }
+
+            this._swap(current, minChild);
+            current = minChild;
+            leftChild = this._getLeftChildIndex(current);
+            rightChild = this._getRightChildIndex(current);
+        }
+
+        return element;
+    }
+
+
+    // The function checks if the heap length is 0.
+    isEmpty() {
+        return this.heap.length === 0;
+    }
+
+    // The function returns the element with the highest priority
+    peek() {
+        return this.heap[0];
+    }
+
+    _getParentIndex(i) {
+        return Math.floor((i - 1) / 2);
+    }
+
+    _getLeftChildIndex(i) {
+        return 2 * i + 1;
+    }
+
+    _getRightChildIndex(i) {
+        return 2 * i + 2;
+    }
+
+    _swap(i, j) {
+        [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
+    }
+}
+
+/**
+ * The class is used to limit the number of concurrent tasks
+ * and the number of tasks executed per minute.
+ * You can use the class to limit the number of requests to a server.
+ * Also, a task in priority queue can be canceled.
+ * @param {number} concurrency - The maximum number of concurrent tasks.
+ * @param {number} rateLimit - The maximum number of tasks executed per minute.
+ * @example
+ * const pool = new AsyncPool(2, 1);
+ * const taskId = await pool.add(async () => {
+ *     await sleep(1000);
+ *     console.log('Task 1');
+ * }, 1);
+ * await pool.add(async () => {
+ *     await sleep(1000);
+ *     console.log('Task 2');
+ * }, 2);
+ * await pool.add(async () => {
+ *     await sleep(1000);
+ *     console.log('Task 3');
+ * }, 3);
+ * pool.cancel(taskId);
```

Comment:
These lines should document the constructor. Just leave the class description above the class.

Reply:
Done